### PR TITLE
Add new helper function get_book_mastery

### PR DESCRIPTION
**Pull request by sacrificerXY**
#### Summary
SUMMARY: Infrastructure "Add new helper function get_book_mastery"

#### Purpose of change

As I was studying how book reading was handled, I noticed some code blocks with this type of pattern.
```
const auto &type = it.type->book;
const skill_id &skill = type->skill;
const int lvl = get_skill_level( skill );
...
if( skill && lvl < type->req ) { ... }
else if ( skill && lvl < type->level ) { ... }
```
This PR aims to introduce a new helper in order to simplify the above code. I didn't change any existing code yet so nothing should be affected.

#### Describe the solution

I added the following enum and a helper function.
```
enum book_mastery {
    CANT_DETERMINE, // book not yet identified, so you don't know yet
    CANT_UNDERSTAND, // does not have enough skill to read
    LEARNING,
    MASTERED // can no longer increase skill by reading
};
book_mastery Character::get_book_mastery( book );
```
Using the example previously mentioned, we can then simplify it like so.
```
book_mastery mastery = get_book_mastery( it );
...
if ( mastery == book_mastery::CANT_UNDERSTAND ) { ... }
else if ( mastery == book_mastery::LEARNING ) { ... }
```

A simple example but as you can see, this new version has less symbols but more English. I'd say this is more readable and thus is an improvement.

#### Describe alternatives you've considered

I considered implementing this as a free function in its own namespace. I decided against it since most functions seems to be attached to classes (from what I've seen).
#### Testing

All tests passed, including the new unit tests I added.

#### Additional context

~~If this gets merged, I'll incorporate them in my next PR to showcase its usage.~~
EDIT: Added commit to demonstrate it's usage.

### CODE REVIEW

File: src/character.cpp [hunk 1]
```diff
@@ -12732,3 +12732,31 @@ int Character::item_reload_cost( const item &it, const item &ammo, int qty ) con
...
 }
+
+book_mastery Character::get_book_mastery( const item &book ) const
+{
+    if( !book.is_book() || !has_identified( book.typeId() ) ) {
+        return book_mastery::CANT_DETERMINE;
+    }
+    // TODO: add illiterate check?
+
+    const auto &type = book.type->book;
```
Comment (anothersimulacrum) [inline src/character.cpp:12743]:
```suggestion
    const islot_book &type = book.type->book;
```

File: tests/reading_test.cpp [hunk 2]
```diff
@@ -336,3 +336,78 @@ TEST_CASE( "reasons for not being able to read", "[reading][reasons]" )
...
+        REQUIRE( alpha.is_book() );
+        REQUIRE_FALSE( dummy.has_identified( alpha.typeId() ) );
+        CHECK( dummy.get_book_mastery( child ) == book_mastery::CANT_DETERMINE );
+    }
+    GIVEN( "an identified book which gives/requires no skill to read" ) {
+        dummy.do_read( child );
+        dummy.do_read( alpha );
+        REQUIRE( dummy.has_identified( child.typeId() ) );
+        REQUIRE( dummy.has_identified( alpha.typeId() ) );
+        const auto &type = child.type->book;
```
Comment (anothersimulacrum) [inline tests/reading_test.cpp:364]:
```suggestion
        const islot_book &type = child.type->book;
```

File: tests/reading_test.cpp [hunk 3]
```diff
@@ -336,3 +336,78 @@ TEST_CASE( "reasons for not being able to read", "[reading][reasons]" )
...
+        const skill_id &skill = type->skill;
+        REQUIRE_FALSE( skill );
+
+        THEN( "you already mastered it" ) {
+            CHECK( dummy.get_book_mastery( child ) == book_mastery::MASTERED );
+        }
+        THEN( "books which give no skill or has no skill requirement are always mastered" ) {
+            const auto &type = child.type->book;
```
Comment (anothersimulacrum) [inline tests/reading_test.cpp:372]:
```suggestion
            const auto &islot_book = child.type->book;
```

File: tests/reading_test.cpp [hunk 4]
```diff
@@ -336,3 +336,78 @@ TEST_CASE( "reasons for not being able to read", "[reading][reasons]" )
...
+        dummy.set_skill_level( skill_id( "chemistry" ), 5 );
+    }
+    GIVEN( "some identified books which gives/requires no skill to read" ) {
+        dummy.do_read( child );
+        dummy.do_read( alpha );
+        REQUIRE( dummy.has_identified( child.typeId() ) );
+        REQUIRE( dummy.has_identified( alpha.typeId() ) );
+
+        WHEN( "it gives/requires no skill" ) {
+            const auto &type = child.type->book;
```
Comment (anothersimulacrum) [inline tests/reading_test.cpp:386]:
```suggestion
            const islot_book &type = child.type->book;
```

File: tests/reading_test.cpp [hunk 5]
```diff
@@ -336,3 +336,78 @@ TEST_CASE( "reasons for not being able to read", "[reading][reasons]" )
...
+            const skill_id &skill = type->skill;
+            REQUIRE_FALSE( skill );
+            THEN( "you already mastered it" ) {
+                CHECK( dummy.get_book_mastery( child ) == book_mastery::MASTERED );
+            }
+
+        }
+        WHEN( "it requires/give skills" ) {
+            const auto &type = alpha.type->book;
```
Comment (anothersimulacrum) [inline tests/reading_test.cpp:395]:
```suggestion
            const islot_book &type = alpha.type->book;
```

**Comment by anothersimulacrum** [comment]
Thanks for adding tests!

File: src/character.cpp (hunk 1 above)
Comment (anothersimulacrum) [reply src/character.cpp]: Ah whoops, I forgot it wasn't just the raw type.

**Comment by kevingranade** [comment]
The code looks good, but it needs to be used at least once in the codebase to be valid to merge.

**Comment by sacrificerXY** [comment]
Added new commit to demonstrate its usage. I could've used it in more places but I'm currently working on another PR which would supersede the changes so... I'll hold off for now.

**Pull request merged**